### ISSUE TITLE
Track length is not considered when calculating average album rating

### ISSUE THREAD

**Issue by bassstorm** [quodlibet/quodlibet#3048]
Simple case: album of two tracks. 1st track is `20` minutes long, rated `0.4`. 2nd track is `4` minutes long, rated `0.6`.
Currently that gives a simple `0.5` average rating for the album. But one would expect that longer piece's rating puts more influence on average (i.e. `(5*0.4 + 1*0.6) / 6 = 0.43` for above case). For example, I don't think last track should affect average [U.F.Orb](https://www.discogs.com/The-Orb-UFOrb/release/13274) album rating much.
Such an improvement in calculation would allow to have more accurate evaluation for albums with short interludes or few very long pieces (see even more... [example 1](https://www.discogs.com/Boards-Of-Canada-Geogaddi/release/24432) or [example 2](https://www.discogs.com/Goldie-Saturnz-Return/release/2349297)).

**Comment by frestr** [comment]
Thanks for the suggestion. This could be added next to the bayesian averaging option in the preferences.

**Comment by lazka** [comment]
We do the same for the bitrate already. I wouldn't mind making it the default.

**Comment by declension** [comment]
Whilst this seems good to me - worth noting that this calculation does penalise [good shorter songs](https://pigeonsandplanes.com/in-depth/2012/10/the-25-best-songs-under-2-minutes) and might get skewed by those [long hidden CD tracks](https://en.wikipedia.org/wiki/List_of_albums_containing_a_hidden_track:_A).

**Comment by CreamyCookie** [comment]
This is a cool idea. 

> worth noting that this calculation does penalise good shorter songs and might get skewed by those long hidden CD tracks.

Good point. If this is implemented it would be best to use a smoothing function to reduce the impact from extremes. Of the top of my head:

```python
>>> from statistics import median, mean
>>> durationInSeconds = [120, 400, 60, 80, 600]
>>> medianDuration = median(durationInSeconds)
>>> medianDuration
120
>>> mean(durationInSeconds)
252                                                                       
>>> SMOOTHING_FACTOR = 0.05                                                                                                                                
>>> smoothed = [i + i * SMOOTHING_FACTOR * (medianDuration - i) / medianDuration for i in durationInSeconds]
>>> smoothed
[120.0, 353.3333333333333, 61.5, 81.33333333333333, 480.0]
>>> # absolute delta between raw and smoothed
>>> [abs(i - s) for i, s in zip(durationInSeconds, smoothed)]
[0.0, 46.666666666666686, 1.5, 1.3333333333333286, 120.0]
```

**Comment by bassstorm** [comment]
Do you guys have a bounty program or something? :) I actually quickly tried to find how to donate, but failed.
These numerous rating improvements (#963, #1510, #1852, #3046, #3426, #3533, #3561, #3605) would make cross-system and library exploration experience much much better.

**Comment by declension** [comment]
> Do you guys have a bounty program or something? :) I actually quickly tried to find how to donate, but failed.

Basically no, other than a few attempts at starting. I think it's safe to say time is the precious commodity here.
Voting on issues (:+1: / :heart:   etc) is something I'd like to encourage more of though, gives a quick way to see what the appetite for things really is 

> These numerous rating improvements (#963, #1510, #1852, #3046, #3426, #3533, #3561, #3605, and infamous--but not rating related--#48 of course) would make cross-system and library exploration experience much much better.

Agreed, slightly depressing when seeing the size of the list though. Also, these types of changes are often high risk to many of the biggest users, so I guess that explains lack of appetite around them

**Comment by bassstorm** [comment]
> Agreed, slightly depressing when seeing the size of the list though.

Yes, and for external people learning curve is probably a bit too high. Even for me as a java developer it is quite hard to follow the code, barely enough for some simple troubleshooting...

> Also, these types of changes are often high risk to many of the biggest users, so I guess that explains lack of appetite around them

True. In fact, having any of them would make us happier :) If you ask me, some of those look like less effort for huge benefit (e.g. #1852, #3533, and #3426 could be likely closed by #3561).
At the end of the day, advanced query mechanism (and rating system as essential part of it) is what makes QL so special!

**Comment by CreamyCookie** [comment]
Okay, I would be interested in implementing this (#3048), so I've created [another test](https://gist.github.com/CreamyCookie/7192358a9f3dbb32cc1ef01784509b3d) roughly showing how it would look:

```
total duration in minutes: 176.5

'smoothed' and 'simple ratio' both represent the ratio / percentage a 
tracks' rating makes up of the whole average rating

rating goes from 0 to 5 (including)

rating    minutes   ➜   smoothed    (rating)  VERSUS  simple ratio    (rating)
5            0.2          0.6%      0.0056              0.1%          0.0009
3            1.7          5.1%      0.0303              0.9%          0.0057
2           16.7         26.4%      0.1056              9.4%          0.0378
2            2.0          5.9%      0.0238              1.1%          0.0045
4            2.7          7.6%      0.0610              1.5%          0.0121
5          150.0         45.2%      0.4524             85.0%          0.8499
1            3.3          9.2%      0.0184              1.9%          0.0038
------------------------------------------------------------------------------
TOTAL:                              0.6971                            0.9146
```

As can be seen, without smoothing small good songs would have little impact on the average rating, while long (possibly hidden tracks as mentioned by @declension) would be given way too much weight.

I'd implement this change for both `avg` and `bav`, and I'd prefer to keep the old way around by adding an additional `basic_avg` (or similar name - suggestions welcome).

Any suggestions / ideas / objections? :smiley:

**Comment by CreamyCookie** [comment]
Okay, I would be interested in implementing this, so I've created [another test](https://gist.github.com/CreamyCookie/7192358a9f3dbb32cc1ef01784509b3d) roughly showing how it would work and look (the `simple ratio` side is how it would look without smoothing):

```
total duration in minutes: 176.5

'smoothed' and 'simple ratio' both represent the ratio / percentage a 
tracks' rating makes up of the whole average rating (i.e. the weight)

rating goes from 0 to 5 (including)

rating    minutes   ➜   smoothed    s_rating  VERSUS  simple ratio    sr_rating
5            0.2          0.6 %     0.0056              0.1 %         0.0009
3            1.7          5.1 %     0.0303              0.9 %         0.0057
2           16.7         26.4 %     0.1056              9.4 %         0.0378
2            2.0          5.9 %     0.0238              1.1 %         0.0045
4            2.7          7.6 %     0.0610              1.5 %         0.0121
5          150.0         45.2 %     0.4524             85.0 %         0.8499
1            3.3          9.2 %     0.0184              1.9 %         0.0038
------------------------------------------------------------------------------
TOTAL:                              0.6971                            0.9146
```

As can be seen, without smoothing small good songs would have little impact on the average rating, while long tracks (possibly hidden ones as mentioned by @declension) would be given way too much weight.

I'd implement this change for both `avg` and `bav`, and I think I'd prefer to keep the old way around by adding an additional `unweighted_avg` (or similar - suggestions welcome). Maybe an `unweighted_bav` as well.. 

Any comments / ideas / objections? :smiley:

**Comment by bassstorm** [comment]
@CreamyCookie sounds very very brilliant! Maybe one little suggestion is to add current state to the table (w/o length).

As a purist, I'd still go with simple ratio (logically it's more correct as you track actual listening time, besides that not so common case with hidden tracks which is almost gone in the digital era) :) But firstly, literally any improvement in that area would be great, and secondly, that may be achievable anyway by tuning `n` value (sorry, math is not my strongest skill). If not, I'm still happy with figures I see above in the table.

P.S. I'm confused why `sr_rating` for 2.7mins is bigger that for 3.3mins?..

**Comment by declension** [comment]
Haven't had a chance to look at the code / stats (sounds good in some way)

Definitely prefer this being optional, or at least like with `bav` to have a prefs item for the length-weighting here (one of which is off)

Naming-wise, IMO  _compressed_  is *so* overloaded for us (byte-compression, like ZIP or FLAC, dynamic range compression, as in high ReplayGain, and the compressor plugins) that it's adding enough confusion that I'd prefer _smoothing_ even if it's wrong(ish), or perhaps there's some other word out there?

**Comment by CreamyCookie** [comment]
@bassstorm 

> Maybe one little suggestion is to add current state to the table (w/o length).


Do you mean adding another column that contains the values of the rating column divided by 5?

> P.S. I'm confused why sr_rating for 2.7mins is bigger that for 3.3mins?..

the values:
```
4            2.7                        1.5 %         0.0121
1            3.3                        1.9 %         0.0038
```

`sr_rating` is the result of `(star_rating / 5) * weight` (since here 5 is the highest possible rating)

for a duration of 2.7 we do `1.5% * (4 / 5) = 0.0121`
for a duration of 3.3 we do `1.9% * (1 / 5) = 0.0038`



> that may be achievable anyway by tuning n value (sorry, math is not my strongest skill). If not, I'm still happy with figures I see above in the table.

Nice! :)

I could set `n` to 10 and make it configurable, as @declension  suggested. I'd do so in the advanced settings, since it seems a bit of niche thing and the normal settings menu is already kinda complicated looking.

The cool thing is that with `n` = 0, the `hyperbolic_smoother` just returns the value as is.

**Comment by CreamyCookie** [comment]
> Naming-wise, IMO compressed is so overloaded for us (byte-compression, like ZIP or FLAC, dynamic range compression, as in high ReplayGain, and the compressor plugins) that it's adding enough confusion that I'd prefer smoothing even if it's wrong(ish), or perhaps there's some other word out there?

Good point! I really can't think of anything, so I guess we'll continue with smoothing.

**Comment by bassstorm** [comment]
> Do you mean adding another column that contains the values of the rating column divided by 5?

Yes, exactly.

> In short, `sr_rating` of the 2.7 minute track is higher because it got a much higher rating (4 times higher).

Oh dear, I should have noted that myself, it's evening... Thanks for explanation!

**Comment by CreamyCookie** [comment]
@bassstorm Here is the new pretty table :1234: :slightly_smiling_face: 

```
The following (duration-based) weights represent the ratio / percentage
a tracks' rating makes up of the album average rating. 

                                                                          simple
                                      smoothed                             ratio
                                      partial             simple         partial
                          smoothed    weighted            ratio         weighted
rating      minutes   ➜   weight      rating    VERSUS    weight          rating
--------------------------------------------------------------------------------
5★ = 1.0       0.2          0.5%      0.0050                0.1%          0.0009
3★ = 0.6       1.7          4.6%      0.0277                0.9%          0.0057
2★ = 0.4      16.7         26.0%      0.1040                9.4%          0.0378
2★ = 0.4       2.0          5.4%      0.0218                1.1%          0.0045
4★ = 0.8       2.7          7.0%      0.0562                1.5%          0.0121
5★ = 1.0     150.0         47.9%      0.4790               85.0%          0.8499
1★ = 0.2       3.3          8.5%      0.0170                1.9%          0.0038
--------------------------------------------------------------------------------
TOTAL:       176.5                    0.7107                              0.9146
```

I really should have used a library to create a markdown table than doing this by hand :man_facepalming:

**Comment by bassstorm** [comment]
@CreamyCookie now with extended description I really convinced I want that smoothing :) Thanks for good explanation.

**Comment by CreamyCookie** [comment]
@bassstorm Awesome to hear! :)

**Comment by bassstorm** [comment]
A few years later...

I've recently discovered a wonderful _foo_sqlite_ component for foobar2000, my daily driver. With power of SQL, I'm able to do things like that and more. Not as user friendly as it could've been done within QL, but works here and now.
This was last puzzle for making my library setup full and complete (+ no issues like infamous #48). Unfortunately, QL has lost the battle.
